server/src/main/java/org/elasticsearch/action/admin/cluster/configuration/TransportAddVotingTombstonesAction.java

```diff
@@ -0,0 +1,178 @@
+
+        clusterService.getMasterService().submitStateUpdateTask("add-voting-tombstones", new ClusterStateUpdateTask() {
+
+            final ClusterStateObserver observer
+                = new ClusterStateObserver(clusterService, request.getTimeout(), logger, threadPool.getThreadContext());
+
+            private Set<DiscoveryNode> resolvedNodes;
+
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                final DiscoveryNodes allNodes = currentState.nodes();
+                assert resolvedNodes == null : resolvedNodes;
+                resolvedNodes = Arrays.stream(allNodes.resolveNodes(request.getNodeDescriptions()))
+                    .map(allNodes::get).filter(DiscoveryNode::isMasterNode).collect(Collectors.toSet());
+
+                if (resolvedNodes.isEmpty()) {
+                    throw new IllegalArgumentException("add voting tombstones request for " + Arrays.asList(request.getNodeDescriptions())
+                        + " matched no master-eligible nodes");
+                }
+
+                resolvedNodes.removeIf(n -> currentState.getVotingTombstones().contains(n));
+                if (resolvedNodes.isEmpty()) {
+                    throw new IllegalArgumentException("add voting tombstones request for " + Arrays.asList(request.getNodeDescriptions())
+                        + " matched no master-eligible nodes that do not already have tombstones");
+                }
+
+                final int oldTombstoneCount = currentState.getVotingTombstones().size();
+                final int newTombstoneCount = resolvedNodes.size();
+                final int maxTombstoneCount = MAXIMUM_VOTING_TOMBSTONES_SETTING.get(currentState.metaData().settings());
+                if (oldTombstoneCount + newTombstoneCount > maxTombstoneCount) {
+                    throw new IllegalArgumentException("add voting tombstones request for " + Arrays.asList(request.getNodeDescriptions())
+                        + " would add [" + newTombstoneCount + "] voting tombstones to the existing [" + oldTombstoneCount
+                        + "] which would exceed the maximum of [" + maxTombstoneCount + "] set by ["
+                        + MAXIMUM_VOTING_TOMBSTONES_SETTING.getKey() + "]");
+                }
+
+                final Builder builder = ClusterState.builder(currentState);
+                resolvedNodes.forEach(builder::addVotingTombstone);
+                final ClusterState newState = builder.build();
+                assert newState.getVotingTombstones().size() <= maxTombstoneCount;
+                return newState;
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+
```

Comment:
can you only start creating the cluster state observer here? I think it's not required as a field.

Reply:
It was here so it could measure the timeout from the time it hits the master node, but that's no longer what the timeout means so I moved it.